### PR TITLE
PoC for removing check results

### PR DESCRIPTION
**Pull request by LucasSte**
#### Problem

SVM's entry point requires a check result for each transaction. For external users, this may be a burden because it may hinder creating a standard behavior for all executed transactions.

#### Summary of Changes

I abstracted the check results in a trait `TransactionCheckResults` that allows a custom implementation to fetch the check result. Such a setting allows external users to customize the behavior or set a default for every transaction in the batch.

This PR is a proof of concept and is not ready to be merged. It is supposed to kickstart a discussion about the solution and possibly devise other changes.

### CODE REVIEW

**Comment by LucasSte** [comment]
I coded this proof of concept for us to discuss about. Can you @buffalojoec @pgarg66 and @dmakarov share your opinion?

**Comment by dmakarov** [comment]
Looks good and nice to me on the first iteration.

**Comment by buffalojoec** [comment]
@LucasSte Thanks for putting this together! I'm wondering what your thoughts might be on an alternative, one which will effectively remove the parameter from the API completely.

Within Bank, an array of `check_results` is created, containing results that describe whether or not:

- The accounts within a transaction were successfully locked.
- The transaction is not expired.
- The transaction was not already processed.

Then, within SVM, the provided check results are immediately zipped alongside the provided transactions to return a list of `validation_results`, and never used again. The returned `validation_results` is yet _another_ list of check results with the same length as the list of transactions. It's used in two spots, and _again_ just zipped alongside the transactions.

Since Bank creates the list of `sanitized_txs` right before invoking the SVM API, we could just make the list of sanitized transactions mutable, iterate over it, and apply the necessary check results directly to each transaction. If we're always going to zip results alongside transactions, it makes sense for the transaction to just contain the result info natively.

https://github.com/anza-xyz/agave/blob/82aeef812c484933947694f21dbe45b4f0505d81/runtime/src/bank.rs#L3468-L3475

We could accomplish this by simply adding a method to the `SVMMessage` trait that vends this result.

```rust
pub struct CheckedTransactionDetails {
    pub nonce: Option<NonceInfo>,
    pub lamports_per_signature: u64,
}

pub trait SVMMessage: Debug {
    fn checked_result(&self) -> transaction::Result<CheckedTransactionDetails> {
        Ok(CheckedTransactionDetails {
            nonce: None,
            lamports_per_signature: 0,
        })
    }

    /* ... */
}
```

We could even further refine it by breaking it up into a few separate methods.

```rust
pub trait SVMMessage: Debug {
    fn pre_processing_result(&self) -> transaction::Result<()> {
        Ok(())
    }

    fn lamports_per_signature(&self) -> u64 {
        0
    }

    fn nonce_info(&self) -> Option<NonceInfo> {
        None
    }

    /* ... */
}
```

With this approach, the result is on the transaction itself, we never have to do any zipping, and we can drop the parameter from the SVM API! What do you think?

---

Side note: Does anyone know why we have `lamports_per_signature` in so many places? Perhaps we should unify this?

File: svm/src/account_loader.rs [hunk 1]
```diff
@@ -39,6 +39,41 @@ pub(crate) type TransactionProgramIndices = Vec<Vec<IndexOfAccount>>;
 pub type TransactionCheckResult = Result<CheckedTransactionDetails>;
 pub type TransactionValidationResult = Result<ValidatedTransactionDetails>;
 
+/// This trait abstract the retrieval of transaction check results for a given transaction
+/// in the transaction batch passed to the SVM.
+pub trait TransactionCheckResults {
+    /// The function consumes a check result and returns it, meaning that once it is requested,
+    /// it is no longer available. `idx` is the index of a transaction in the batch.
+    fn consume_check_result(&mut self, _idx: usize) -> TransactionCheckResult {
```
Comment (apfitzge) [inline svm/src/account_loader.rs:47]: it looks like the first thing we do during transaction iteration is call this consume fn.
If that's the case, why not have either an iterator return, or an `IntoIterator` bound?

It seems simple enough to have some dummy struct that impls this with a repeating iterator to get similar behavior, but also doesn't eat (minor) cost of replacement for the vec impl.

**Comment by apfitzge** [comment]
> We could accomplish this by simply adding a method to the SVMMessage trait that vends this result.

@buffalojoec If doing this, **please** create an extension trait like `CheckedSVMTransaction`.
Otherwise I'll end up creating a trait identical to what `SVMTransaction` is now that doesn't have that stuff.

**Comment by buffalojoec** [comment]
> > We could accomplish this by simply adding a method to the SVMMessage trait that vends this result.
> 
> @buffalojoec If doing this, **please** create an extension trait like `CheckedSVMTransaction`. Otherwise I'll end up creating a trait identical to what `SVMTransaction` is now that doesn't have that stuff.

Yeahhh I figured it wouldn't be the most ideal thing to start overloading that trait. But agreed another trait works as well.

**Comment by LucasSte** [comment]
@buffalojoec That is a good approach. At first, I wanted to avoid cloning `NonceInfo` all the time, so I decided to take the other approach. Let me see if I can make the sanitized transaction mutable.

**Comment by LucasSte** [comment]
@buffalojoec I don't think it is possible to make `SanitizedTransaction` mutable, because `check_transactions` receive a `&[impl core::borrow::Borrow<SanitizedTransaction>]`. One of its callers is indeed passing a slice of references `&[&SanitizedTransaction]`.

https://github.com/anza-xyz/agave/blob/4a777d1288ec7d935db8d8d45c3eff16a4ade0a3/core/src/banking_stage/transaction_scheduler/scheduler_controller.rs#L224-L232

We can alternatively use a `RefCell` to make the check results mutable, however I think it is a bit ugly.

**Comment by LucasSte** [comment]
Alternatively, we could create another wrapper for SanitizedTransaction in the SVM, which would hold both the transactions and the check results.

**Comment by LucasSte** [comment]
> Since Bank creates the list of sanitized_txs right before invoking the SVM API, we could just make the list of sanitized transactions mutable, iterate over it, and apply the necessary check results directly to each transaction. If we're always going to zip results alongside transactions, it makes sense for the transaction to just contain the result info natively.

Bank retrieves a reference to a slice from `TransactionBatch`, which holds a `Cow<[SanitizedTransaction]>`. Everything is always created from references. Up from ledger's `TransactionBatchWithIndexes`. Making the list mutable is going to require changes in too many places.

**Comment by buffalojoec** [comment]
> Bank retrieves a reference to a slice from `TransactionBatch`, which holds a `Cow<[SanitizedTransaction]>`. Everything is always created from references. Up from ledger's `TransactionBatchWithIndexes`. Making the list mutable is going to require changes in too many places.

Hmm yeah looks like it might be more trouble than it's worth. Your other suggestion is still aligned with what I was leaning toward though!

> Alternatively, we could create another wrapper for SanitizedTransaction in the SVM, which would hold both the transactions and the check results.

It sounds like it wouldn't be super pretty, but basically some struct `TransactionWithCheckResult` that can hang onto the original reference for the tx but also store a check result. If we use a slice of transactions to make this new slice, perhaps we don't even need anything to be mutable at all. Does that sound right?

The concern I have though is that we're going to need SVM to call this new method somewhere, and for that reason it makes sense to have it live under some trait dubbed "SVM-" something. Making it a separate adjacent trait might confuse people. We might have to bikeshed the trait specifics, but at least with this approach you get rid of this awkward input parameter to the API.

**Comment by LucasSte** [comment]
> It sounds like it wouldn't be super pretty, but basically some struct TransactionWithCheckResult that can hang onto the original reference for the tx but also store a check result. If we use a slice of transactions to make this new slice, perhaps we don't even need anything to be mutable at all. Does that sound right?

Yes, that is correct. I was thinking of having a `struct TransactionWithCheckResult` which holds both the check results and a reference to the slice of transaction. You can create a trait to return the NonceInfo and the lamports per signature.

This way, we can call `load_and_execute_transactions` with the new struct. How does this sound?

I had also thought of wrapping each one of the sanitized transaction with a struct to holds its respective check result, but that would require iterating over the returned results. I am not so keen on adding another pass over the vector.

**Comment by buffalojoec** [comment]
> Yes, that is correct. I was thinking of having a `struct TransactionWithCheckResult` which holds both the check results and a reference to the slice of transaction. You can create a trait to return the NonceInfo and the lamports per signature.
> 
> This way, we can call `load_and_execute_transactions` with the new struct. How does this sound?

So `load_and_execute_transactions` will require a struct that allows implementations of both `SVMTransaction` and this new trait proposed within it? Something like `struct Joe<A, B> where A: SVMMessage, B: SVMCheckResult`?

**Comment by apfitzge** [comment]
> So load_and_execute_transactions will require a struct that allows implementations of both SVMTransaction and this new trait proposed within it? Something like struct Joe<A, B> where A: SVMMessage, B: SVMCheckResult?

Maybe simpler to just have it as child trait (not sure correct naming)?

i.e. `pub trait CheckedSVMTransaction: SVMTransaction {}`

this way the method can just take `CheckedSVMTransaction`, and it will still have access to all the methods of `SVMTransaction`.

The type itself could just wrap an `Tx: SVMTransaction` with the check result.
The annoying thing is that we have a bunch of boiler plate in re-impling (really just routing the function calls) of the wrapper type.

**Comment by buffalojoec** [comment]
> > So load_and_execute_transactions will require a struct that allows implementations of both SVMTransaction and this new trait proposed within it? Something like struct Joe<A, B> where A: SVMMessage, B: SVMCheckResult?
> 
> Maybe simpler to just have it as child trait (not sure correct naming)?
> 
> i.e. `pub trait CheckedSVMTransaction: SVMTransaction {}`
> 
> this way the method can just take `CheckedSVMTransaction`, and it will still have access to all the methods of `SVMTransaction`.
> 
> The type itself could just wrap an `Tx: SVMTransaction` with the check result. The annoying thing is that we have a bunch of boiler plate in re-impling (really just routing the function calls) of the wrapper type.

You could maybe avoid the routing by just doing `trait CheckedSVMTransaction<T: SVMTransaction> {}`. Not sure if the generic parameter makes this annoying, though, since the API would then need to shift `SVMTransaction` to a generic arg, ie. `process_txs<CB: Callbacks, T: SVMTransaction>(&self, callbacks: &CB, txs: &[CheckedSVMTransaction<T>])`.

**Comment by LucasSte** [comment]
I believe we can avoid the extra parameters by doing this (I haven't tested it yet):

```Rust
struct CheckedSVMTransaction<'a, 'b> {
    transactions: &'a [impl SVMTransaction],
    checked_results: &'b [impl SVMCheckedResults],
}
```

The struct size is constant because it only holds slices.

**Comment by LucasSte** [comment]
Or better:

```
struct CheckedSVMTransaction<'a, 'b> {
    transactions: &'a [impl SVMTransaction],
    checked_results: Vec<impl SVMCheckedResults>,
}
```

**Comment by buffalojoec** [comment]
> Or better:
> 
> ```
> struct CheckedSVMTransaction<'a, 'b> {
>     transactions: &'a [impl SVMTransaction],
>     checked_results: Vec<impl SVMCheckedResults>,
> }
> ```

How is this different than the current implementation, though?

```rust
process_ixs(
    transactions: &[impl SVMTransaction],
    checked_results: &[impl SVMCheckedResults],
)
```

**Comment by LucasSte** [comment]
> How is this different than the current implementation, though?

I had to read through all the messages again to understand the discussion. I thought we were talking about the same constructions, and agreeing on them. My wrapper suggestion was to encapsulate both lists since the beginning. I agree that it is not a strong departure from what is implemented in this PR.

One of the variations I had thought of is the following. The trait could instead be implemented for SVMCheckedTransaction, allowing us to return a zipped iterator for both lists when checks are present and dummy values and the checks are not there.

```rust
struct SVMCheckedTransaction<'a, 'b> {
      txs: &'a [impl SVMTransaction],
      checks: Optional<&b [TransactionCheckResults],
}
```

I wasn't so keen on having a list of a struct to encapsulate each element of SVMTransaction and CheckedResults. That is because the creation of such a structure entails an iteration over the transactions:

> I had also thought of wrapping each one of the sanitized transaction with a struct to holds its respective check result, but that would require iterating over the returned results. I am not so keen on adding another pass over the vector.

I also agree with @apfitzge that such an approach would require a lot of repeated code to reimplement the `trait SVMTransaction`. That would increase the maintenance burden.

**Comment by buffalojoec** [comment]
@LucasSte there's also another option of internalizing the creation of the `check_results` list within SVM, and instead using a callback hook on `callbacks: &CB` that does something like `check_transaction`. Then Bank can implement this method, and the parameter comes out of the API. What do you think?

Additionally, some of the extra zipping and transforming of the check results could even be combined with this approach, simplifying the internals a bit.

**Comment by LucasSte** [comment]
> @LucasSte there's also another option of internalizing the creation of the `check_results` list within SVM, and instead using a callback hook on `callbacks: &CB` that does something like `check_transaction`. Then Bank can implement this method, and the parameter comes out of the API. What do you think?
> 
> Additionally, some of the extra zipping and transforming of the check results could even be combined with this approach, simplifying the internals a bit.

That is a smart alternative. I'll implement it and open a new PR.

**Comment by LucasSte** [comment]
`check_transactions` requires `lock_results` as a parameter. If we internalize the `fn check_transactions` to the SVM, we'll need this extra parameter. In this case, we end up in the same initial problem again.

https://github.com/anza-xyz/agave/blob/b345960389581f47b8018062241c811079299702/runtime/src/bank/check_transactions.rs#L56

We may consider then passing `struct TransactionBatch` to the SVM, which would require some refactoring around runtime to properly remove its brank reference.

**Comment by apfitzge** [comment]
> We may consider then passing struct TransactionBatch to the SVM

As someone who has recently run into a bunch of trouble with `TransactionBatch`, please don't.

**Comment by LucasSte** [comment]
Changing the SVM API to process single transactions isn't happening soon, although I believe we are going to do it at some point.

The approach I took in this PR is a bit clumsy in that I could return an iterator in `consume_check_result` (possibly renaming the function too). However, my solution only slightly simplifies the problem and we still need to mock a check result for external usages of SVM, just like creating a vector as we do now.

I couldn't see any other straightforward approach to solve the problem, without requiring major changes everywhere. I am leaning towards waiting until we have a streaming API to fix the problem, instead of pursuing this change.

Any thoughts on that?

**Pull request closed without merging**